aten/src/TH/generic/THStorageCopy.cpp

```diff
@@ -5,49 +5,61 @@
 void THStorage_(rawCopy)(THStorage *storage, real *src)
 {
   ptrdiff_t i;
+  real *data = THStorage_(data)(storage);
   for(i = 0; i < storage->size; i++)
-    storage->data[i] = src[i];
+    data[i] = src[i];
 }
 
 void THStorage_(copy)(THStorage *storage, THStorage *src)
 {
   THArgCheck(storage->size == src->size, 2, "size mismatch");
-  THStorage_(rawCopy)(storage, src->data);
+  THStorage_(rawCopy)(storage, THStorage_(data)(src));
 }
 
+// NOTE: for performance, these macros generally directly access data, rather
+// than THStorage_(data)ß
+
 #define IMPLEMENT_THStorage_COPY(TYPENAMESRC) \
 void THStorage_(copy##TYPENAMESRC)(THStorage *storage, TH##TYPENAMESRC##Storage *src) \
 { \
-  ptrdiff_t i;                                                        \
-  for(i = 0; i < storage->size; i++)                                  \
-    storage->data[i] = (real)src->data[i];                            \
+  ptrdiff_t i;                                                          \
+  auto data = THStorage_(data)(storage);                                \
```

Comment:
The use of THStorage_(data) here seems to contradict the Note above.

Reply:
I don't agree, I pull out the data pointers with type checking once and then use the raw pointers directly in the loop.

Reply:
I agree the comment above should be clearer though.